sequence_mining/pboost-1.0/man/pspan.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\"
.TH PSPAN 1 "MAY 2007" pspan "User Manual"
.SH NAME
pspan \- PrefixSpan frequent subsequence mining.
.SH SYNOPSIS
.B pspan
.I [options]
.I input.txt
.SH DESCRIPTION
Mine a set of sequences for frequently appearing subsequences.  Each returned
sequence appears as subsequence in the training set "frequently", its' support
is above a given threshold.

There are two modes of operation: either the minimum required support is given
as absolute value or fraction of the number of training samples (--minsup and
--minsuppct options, respectively), or a constant K is provided and the top K
frequent subsequences are returned (--topk option).

Additionally, a maximum gap constraint can be specified using the --maxgap
option (see the documentation of the
.BR pboost
program for the meaning of the maximum gap constraint).

The resulting set of sequences can be written to an output file and projection
of the training set onto these sequences can be carried out (--project
option).

.SH GENERIC OPTIONS
.IP "--help"
Print a short usage summary.
.SH PARAMETERS
.IP "--output <file.txt>"
Set the output file name where the found subsequences will be written to.
By default this is set to "output.txt".
.IP "--project <file.txt>"
After mining the frequent subsequences, project the training data onto the
found sequences, producing for n training sequences and k frequent
subsequences an n-times-k binary matrix encoding which subsequence appears in
which training sequence.
.IP "--verbose"
Enable verbose mode.
.IP "--verify"
Debug option: explicit subsequence occurence verification.  Because PrefixSpan
implicitly enumerates the sequences, the meta-information about each found
sequence is the result of a long chain of computations.  This option enables
an explicit verification function that verifies the meta-information for each
found sequence and abort the program if an inconsitent state is found.  This
should be used only for debugging purposes as it slows down the program
considerably.
.IP "--topk <value>"
Mine the top-K frequent subsequences.  The mining is carried out using the
A-Star algorithm and is very efficient.

At least K subsequences are returned, usually more.  The reason is that given
a certain threshold K there might be more sequences equally frequent as the
K'th found sequence.  Such sequences are equally frequent and hence equally
important.

This option is the default and set to 20.
.IP "--minsup <value>"
Explicitly give the minimum required support as absolute sequence count.
.IP "--minsuppct <value>"
Give the minimum required support as fraction of the training sample count.
Thus, 0.0 < value <= 1.0 must hold.
.IP "--length-min <value>"
Minimum required subsequence length, measured in total item count.  The
default is 1.
.IP "--length-max <value>"
Maximum allowed subsequence length, measured in total item count.  If zero (0)
is given, no maximum length is enforced.  The default is zero.
.IP "--maxgap <value>"
Maximum allowed element gap when matching, or -1 for infinite gaps.  The
default is -1.
.SH SEQUENCE DATA
A sequence is an ordered list of one or more elements.  An element is an
set of zero or more items.  An item is a positive integer number.

The subsequence relationship is defined between two sequences s1 and s2 as:
s1 is a subsequence of s2, if and only if a monotonically-increasing index
mapping I for each element in s1 can be established, such that each element
s1(k) is a subset of s2(I(k)).  That is, a sequence is a subsequence of
another, if it can be matched with arbitrary long gaps but preserving the
order, such that the matched elements satisfy a subset relationship.

If a maximum gap constraint is used, the above holds but additionally the
difference of each pair of adjacent indices in the mapping I must be no
greater then maxgap+1.
.SH PERFORMANCE
On a medium sized test problem (383 sequences, with 15 elements and an average
item count of circa 400) the following timings have been obtained, extracting
the top-500 frequent subsequences:
1m10s for infinite gap length (-G -1), for maximum gap constrained problems:
1m54s (-G 0), 3m00s (-G 1), 4m20s (-G 2), 5m55s (-G 3), 7m35s (-G 4) and
9m13s (-G 5).

This highlights the relative speeds of mining frequent subsequences with
arbitrary long gaps (fastest) compared to maximum-gap constrained subsequences
with high gap lengths (slowest).  The performance is empirically linear in the
maximum allowed gap length.

When using the iterative deepening A-Star search strategy, memory usage should
never be a problem, even for a larger database of sequences.
.SH INPUT FILE FORMAT
The input file consists of n lines, each containing a filename.
The filename corresponds to a sequence file.

Each sequence file consists of p lines, each line representing one sequence
element.  Each element contains a number of items (positive integer
numbers), separated by whitespace characters.  The items within each element
must be ordered and contain no duplicate item.
.SH BUGS
No bugs known, if you find any, please send a bug report to me. I will try to
fix it.
.SH AUTHOR
Sebastian Nowozin <sebastian.nowozin@tuebingen.mpg.de>
.SH "SEE ALSO"
.BR pboost (1),
.BR ptest (1)


sequence_mining/pboost-1.0/src/pspan.cpp
/* pspan Subsequence Mining
 * Copyright (C) 2007  Sebastian Nowozin
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

/* pspan.cpp - PrefixSpan frequent subsequence mining
 *
 * Author: Sebastian Nowozin <sebastian.nowozin@tuebingen.mpg.de>
 * Date: 4th July 2007
 */

#include <stdlib.h>
#include <iostream>
#include <fstream>
#include <iterator>

#include <boost/program_options.hpp>

#include <PrefixSpan.h>
#include <PrefixSpanK.h>
#include <PrefixSpanDSPCA.h>

namespace po = boost::program_options;

using namespace SeqMine;


static bool
parseSequenceMetaFile (const char* filename, std::vector<const SeqMine::Sequence*>& S)
{
	std::ifstream in (filename);
	
	/* Read in filenames
	 */
	S.clear ();
	


	for (unsigned int seqId = 0 ; in.eof () == false ; ++seqId) {
		char seqFilename[256];

		in.width (sizeof (seqFilename) - 1);
		seqFilename[0] = '\0';
		in >> seqFilename;
		if (strlen (seqFilename) == 0)
			break;

		std::cout << "  reading \"" << seqFilename << "\"" << std::endl;
		std::cout << "done 1" << std::endl;
		SequenceT* seq = SequenceT::readFromFile (seqFilename);
		std::cout << "done 1" << std::endl;
		seq->Occurence.insert (seqId);
		std::cout << "done 1" << std::endl;
		S.push_back (seq);
	}
	in.close ();

	std::cout << "Successfully read in " << S.size ()
		<< " sequence files." << std::endl;

	return (true);
}

int
main (int argc, char* argv[])
{
	/* Parse command line options
	 */
	po::options_description generic ("Generic options");
	generic.add_options ()
		("help", "Produce help message")    
		;

	/* Output file
	 */
	std::string outputFilename;
	std::string projectFilename;
	unsigned int topK;
	unsigned int dspcaBasisNonZero;
	unsigned int dspcaBasisCount;
	unsigned int minsup = 0;
	double minsupPercent;
	unsigned int length_min, length_max;
	int maxgap;

	po::options_description config ("Parameters");
	config.add_options ()
		("output", po::value<std::string>(&outputFilename)->default_value ("output.txt"),
			"Frequent subsequence output file")
		("project,p", po::value<std::string>(&projectFilename),
			"Project samples on found sequences.")
		("lexicon,L", "Lexicon Construction Mode")
		("verbose,v", "Verbose mode, dumping found subsequences.")
		("verify,V", "Verify all projections explictly (slow).")
		("topk,K", po::value<unsigned int>(&topK)->default_value (20),
			"Mine the top K frequent subsequences.")
		("dspca", po::value<unsigned int>(&dspcaBasisNonZero),
			"Perform direct sparse PCA with the given number of "
			"nonzero coefficients per basis vector")
		("dspcacount", po::value<unsigned int>(&dspcaBasisCount)->default_value (5),
			"The number of PCA basis vectors to extract")
		("minsup,S", po::value<unsigned int>(&minsup), "Minimum support (count)")
		("minsuppct,s", po::value<double>(&minsupPercent),
			"Minimum support tau, 0<tau<=1 (percentile)")
		("length-min",
			po::value<unsigned int>(&length_min)->default_value (1),
			"Minimum pattern sequence length")
		("length-max",
			po::value<unsigned int>(&length_max)->default_value (0),
			"Maximum pattern sequence length or zero for no maximum")
		("maxgap,G", po::value<int>(&maxgap)->default_value (-1),
			"Maximum allowed gap or -1 for infinite gaps")
		;

	po::options_description hidden ("Hidden options");
	hidden.add_options ()
		("input", po::value<std::vector<std::string> >(), "Sequence input file")
		;

	po::options_description visible_options;
	visible_options.add (generic).add (config);

	po::options_description all_options;
	all_options.add (config).add (generic).add (hidden);

	po::positional_options_description p;
	p.add ("input", -1);

	/* Parse
	 */
	po::variables_map vm;
	po::store (po::command_line_parser (argc, argv).options (all_options).positional (p).run (), vm);
	po::notify (vm);

	if (vm.count ("help") || vm.count ("input") == 0 ||
		vm["input"].as<std::vector<std::string> >().size () != 1)
	{
		std::cerr << "pspan, $Id: pspan.cpp + lexicon builder  $" << std::endl;
		std::cerr << "================================================================================" << std::endl;
		std::cerr << "Copyright (C) 2007 -- Sebastian Nowozin <sebastian.nowozin@tuebingen.mpg.de>" << std::endl;
		std::cerr << std::endl;
		std::cerr << "Usage: pspan [options] input.txt" << std::endl;
		std::cerr << std::endl;
		std::cerr << "Required parameters:" << std::endl;
		std::cerr << "  input.txt           List of input sequences." << std::endl;
		std::cerr << visible_options << std::endl;
		std::cerr << std::endl;

		exit (EXIT_FAILURE);
	}

	/* Read sequence files
	 */
	std::vector<std::string> input = vm["input"].as<std::vector<std::string> >();
	std::vector<const SeqMine::Sequence*> S;

	bool seqresult = parseSequenceMetaFile (input[0].c_str (), S);
	if (seqresult == false) {
		std::cerr << "Error while reading in sequence files." << std::endl;

		exit (EXIT_FAILURE);
	}

    /* Lexicon Contruction Code 
        Dustin Smith 2010-04-24 

    */
    if (vm.count("lexicon")) {

        std::cout << "Constructing a lexicon...\n";

        
        int it = 0;
        while(1) {
            
    
            std::cout << "Lexicon construction Iteration " << it << endl;
            
            /* min support percentage mode */
            double minsupPercent = 0.2;
            minsup = (unsigned int) (minsupPercent * ((double) S.size ()) + 0.5);
    		std::cerr << "Minimum support is " << minsupPercent << ", set to "	<< minsup << std::endl;
            
            std::vector<std::pair<unsigned int, SequenceT> > subseq;
        	PrefixSpanOptions poptions (minsup, length_min, length_max, maxgap);
			PrefixSpanK pspan (topK, poptions);
			if (vm.count ("verify"))
				pspan.SetVerification (true);

			pspan.Mine (S);

		    std::cout << "Mined " << subseq.size () << " sequences." << std::endl;
        	std::cout << "The minimum support threshold for K=" << topK	<< " is " << pspan.MinimumSupportThreshold () << std::endl;

			std::copy (pspan.mostFrequent.begin (), pspan.mostFrequent.end (),
				std::back_insert_iterator<std::vector<std::pair<unsigned int, SequenceT> > >
					(subseq));
	
        
        					
    		std::ofstream ssout (outputFilename.c_str ());
    		for (std::vector<std::pair<unsigned int, SequenceT> >::iterator
    			iv = subseq.begin () ; iv != subseq.end () ; ++iv)
    		{
    			ssout << iv->second << std::endl;
    		}
    		ssout.close ();
    
        	/* Dump found sequences
        	 */
        	if (vm.count ("verbose")) {
        		for (std::vector<std::pair<unsigned int, SequenceT> >::iterator
        			iv = subseq.begin () ; iv != subseq.end () ; ++iv)
        		{
        			std::cout << iv->first << " times: " << iv->second << std::endl;
        		}
        	}

        	/* Do projection on subsequences and output file.
        	 */
        	if (vm.count ("project")) {
        		std::ofstream projout (projectFilename.c_str ());

        		for (unsigned int n = 0 ; n < S.size () ; ++n) {
        			for (std::vector<std::pair<unsigned int, SequenceT> >::iterator
        				iv = subseq.begin () ; iv != subseq.end () ; ++iv)
        			{
        				if (iv != subseq.begin ())
        					projout << " ";

        				if (iv->second.Occurence.find (n) != iv->second.Occurence.end ())
        					projout << "1";
        				else
        					projout << "0";
        			}
        			projout << std::endl;
        		}
        		projout.close ();
        	}

    	/* Cleanup
    	 */
    	for (std::vector<const Sequence*>::iterator si = S.begin () ;
    		si != S.end () ; ++si)
    	{
    		delete (*si);
    	}
    	
    	
                break; // exit loop
        } // end while
        
    } else {
    
    	/* Get mode and calculate details
    	 */
    	if (vm.count ("minsup") && vm.count ("minsuppct")) {
    		std::cerr << "The \"minsup\" and \"minsuppct\" options "
    			"are mutually exclusive." << std::endl;

    		exit (EXIT_FAILURE);
    	}

    	bool minsup_mode = false;
    	if (vm.count ("minsup") || vm.count ("minsuppct"))
    		minsup_mode = true;

    	if (vm.count ("minsuppct")) {
    		minsup = (unsigned int) (minsupPercent * ((double) S.size ()) + 0.5);
    		std::cerr << "Minimum support is " << minsupPercent << ", set to "
    			<< minsup << std::endl;
    	}

    	std::vector<std::pair<unsigned int, SequenceT> > subseq;
    	PrefixSpanOptions poptions (minsup, length_min, length_max, maxgap);

    	if (vm.count ("dspca") == 0) {
    		if (minsup_mode) {
    			PrefixSpan pspan (poptions);
    			if (vm.count ("verify"))
    				pspan.SetVerification (true);

    			pspan.Mine (S);

    			std::copy (pspan.subsequences.begin (), pspan.subsequences.end (),
    				std::back_insert_iterator<std::vector<std::pair<unsigned int, SequenceT> > >
    					(subseq));
    		} else if (vm.count ("topk")) {
    			PrefixSpanK pspan (topK, poptions);
    			if (vm.count ("verify"))
    				pspan.SetVerification (true);

    			pspan.Mine (S);

    			std::cout << "The minimum support threshold for K=" << topK
    				<< " is " << pspan.MinimumSupportThreshold () << std::endl;

    			std::copy (pspan.mostFrequent.begin (), pspan.mostFrequent.end (),
    				std::back_insert_iterator<std::vector<std::pair<unsigned int, SequenceT> > >
    					(subseq));
    		}
    	} else {
    		PrefixSpanDSPCA pspan (dspcaBasisNonZero, dspcaBasisCount, poptions);

    		pspan.PCA (S);

    		/* Do PCA projection (different from normal projection):
    		 *
    		 * Xred = (X-repmat(cmean,size(X,1),1))*comp';
    		 */
    		if (vm.count ("project")) {
    			std::ofstream projout (projectFilename.c_str ());

    			for (unsigned int n = 0 ; n < S.size () ; ++n) {
    				for (unsigned int c = 0 ; c < pspan.components ; ++c) {
    					double f = 0.0;

    					for (unsigned int el = 0 ; el < pspan.Cpatterns[c].size () ; ++el) {
    						bool contains = (pspan.Cpatterns[c][el].Occurence.find (n) !=
    							pspan.Cpatterns[c][el].Occurence.end ());

    						f += pspan.Ceval[c][el] * ((contains ? 1.0 : 0.0) -
    							((double) pspan.Cpatterns[c][el].Occurence.size ()) /
    								((double) S.size ()));
    					}

    					if (c > 0)
    						projout << " ";

    					projout << f;
    				}
    				projout << std::endl;
    			}
    			projout.close ();
    		}

    		exit (EXIT_SUCCESS);
    	}

    	std::cout << "Mined " << subseq.size () << " sequences." << std::endl;

    	/* Write output file
    	 */
    	{
    		std::ofstream ssout (outputFilename.c_str ());
    		for (std::vector<std::pair<unsigned int, SequenceT> >::iterator
    			iv = subseq.begin () ; iv != subseq.end () ; ++iv)
    		{
    			ssout << iv->second << std::endl;
    		}
    		ssout.close ();
    	}

    	/* Dump found sequences
    	 */
    	if (vm.count ("verbose")) {
    		for (std::vector<std::pair<unsigned int, SequenceT> >::iterator
    			iv = subseq.begin () ; iv != subseq.end () ; ++iv)
    		{
    			std::cout << iv->first << " times: " << iv->second << std::endl;
    		}
    	}

    	/* Do projection on subsequences and output file.
    	 */
    	if (vm.count ("project")) {
    		std::ofstream projout (projectFilename.c_str ());

    		for (unsigned int n = 0 ; n < S.size () ; ++n) {
    			for (std::vector<std::pair<unsigned int, SequenceT> >::iterator
    				iv = subseq.begin () ; iv != subseq.end () ; ++iv)
    			{
    				if (iv != subseq.begin ())
    					projout << " ";

    				if (iv->second.Occurence.find (n) != iv->second.Occurence.end ())
    					projout << "1";
    				else
    					projout << "0";
    			}
    			projout << std::endl;
    		}
    		projout.close ();
    	}

    	/* Cleanup
    	 */
    	for (std::vector<const Sequence*>::iterator si = S.begin () ;
    		si != S.end () ; ++si)
    	{
    		delete (*si);
    	}
    }
}


